### PR TITLE
CAMEL-20355: switch throttler using Enum/flags

### PR DESCRIPTION
**Pull request by orpiske**
Here's another try.

This one uses an Enum to switch between throttling modes. 

- Known problems: the ManagedThrottlerMBean is not being registered and is not working at the moment. 

I'm still testing this on my own CI, but given it's not so radically different than previous ones, I'm hopeful it will be OK.  

Obs.: the commits are dirty (I know).

### CODE REVIEW

File: catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/models.properties [hunk 1]
```diff
@@ -204,6 +205,7 @@ to
 toD
 tokenize
 topic
+totalRequestsConfig
```
Comment (davsclaus) [inline catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/models.properties:208]: I assume these are left over from the old code

File: catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/models.properties [hunk 2]
```diff
@@ -18,6 +18,7 @@ circuitBreaker
 claimCheck
 combinedServiceDiscovery
 combinedServiceFilter
+concurrentConfig
```
Comment (davsclaus) [inline catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/models.properties:21]: I assume these are left over from the old code

File: catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/models/concurrentConfig.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (davsclaus) [inline catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/models/concurrentConfig.json:1]: I assume these are left over from the old code

File: catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/models/totalRequestsConfig.json [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (davsclaus) [inline catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/models/totalRequestsConfig.json:1]: I assume these are left over from the old code

File: core/camel-core-model/src/main/java/org/apache/camel/model/ThrottleDefinition.java [hunk 5]
```diff
@@ -57,29 +57,60 @@ public class ThrottleDefinition extends ExpressionNode implements ExecutorServic
     @XmlAttribute
     @Metadata(label = "advanced", javaType = "java.lang.Boolean")
     private String rejectExecution;
+    @XmlAttribute
+    @Metadata(defaultValue = "1000", javaType = "java.time.Duration")
+    private String timePeriodMillis;
+
+    @XmlAttribute
+    @Metadata(label = "advanced", javaType = "org.apache.camel.model.ThrottlingMode", defaultValue = "TotalRequests",
```
Comment (davsclaus) [inline core/camel-core-model/src/main/java/org/apache/camel/model/ThrottleDefinition.java:65]: I dont think it should be advanced, and I think this should be the first option, eg move to top so user select mode first.

File: core/camel-core-model/src/main/java/org/apache/camel/model/ThrottleDefinition.java [hunk 6]
```diff
@@ -89,35 +120,132 @@ public String getShortName() {
...
     }
 
+    /**
+     * Sets the maximum number of concurrent requests
+     *
+     * @param      maximumConcurrentRequests the maximum number of concurrent requests
+     * @deprecated                           Use {@link #maximumRequests(long)}
+     * @return                               the builder
+     */
+    @Deprecated(since = "4.4.0")
```
Comment (davsclaus) [inline core/camel-core-model/src/main/java/org/apache/camel/model/ThrottleDefinition.java:182]: The new throttler code was added in 4.3 (non LTS) so IMHO its fine to remove/change the code as upgrade from LTS to LTS is what matters.

File: core/camel-core-model/src/main/java/org/apache/camel/model/ThrottleDefinition.java [hunk 7]
```diff
@@ -89,35 +120,132 @@ public String getShortName() {
...
+
+    /**
+     * Sets the number of concurrent requests
+     *
+     * @param      maximumConcurrentRequests the maximum number of concurrent requests
+     * @deprecated                           Use {@link #maximumRequests(long)}
+     * @return                               the builder
+     */
+    @Deprecated(since = "4.4.0")
+    public ThrottleDefinition maximumConcurrentRequests(String maximumConcurrentRequests) {
```
Comment (davsclaus) [inline core/camel-core-model/src/main/java/org/apache/camel/model/ThrottleDefinition.java:203]: The new throttler code was added in 4.3 (non LTS) so IMHO its fine to remove/change the code as upgrade from LTS to LTS is what matters.

File: core/camel-core-model/src/main/java/org/apache/camel/model/ThrottleDefinition.java [hunk 8]
```diff
@@ -247,6 +375,19 @@ public ThrottleDefinition executorService(String executorService) {
         return this;
     }
 
+    /**
+     * Sets the throttling mode
```
Comment (davsclaus) [inline core/camel-core-model/src/main/java/org/apache/camel/model/ThrottleDefinition.java:379]: We may want a little bit more doc here. Maybe talk a bit what each mode does.

File: core/camel-core-reifier/src/main/java/org/apache/camel/reifier/ThrottleReifier.java [hunk 9]
```diff
@@ -33,30 +35,45 @@ public ThrottleReifier(Route route, ProcessorDefinition<?> definition) {
...
         Expression maxRequestsExpression = createMaxRequestsPerPeriodExpression();
         if (maxRequestsExpression == null) {
             throw new IllegalArgumentException("MaxRequestsPerPeriod expression must be provided on " + this);
         }
 
-        Expression correlation = null;
-        if (definition.getCorrelationExpression() != null) {
-            correlation = createExpression(definition.getCorrelationExpression());
-        }
+        if (ThrottlingMode.toMode(definition.getMode()) == ThrottlingMode.ConcurrentRequests) {
```
Comment (davsclaus) [inline core/camel-core-reifier/src/main/java/org/apache/camel/reifier/ThrottleReifier.java:55]: Add parseString(definition...) as that does property placeholders

File: core/camel-management-api/src/main/java/org/apache/camel/api/management/mbean/ManagedConcurrentRequestsThrottlerMBean.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.api.management.mbean;
+
+import org.apache.camel.api.management.ManagedAttribute;
+
+public interface ManagedConcurrentRequestsThrottlerMBean extends ManagedProcessorMBean {
```
Comment (davsclaus) [inline core/camel-management-api/src/main/java/org/apache/camel/api/management/mbean/ManagedConcurrentRequestsThrottlerMBean.java:21]: Can we avoid having different mbeans - as there is only 1 EIP and ideally we should match that with the mbean too.

Keep the old mbean name, and then add attribute for mode, so you can see there which mode it uses.

File: catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/models.properties (hunk 2 above)
Comment (orpiske) [reply catalog/camel-catalog/src/generated/resources/org/apache/camel/catalog/models.properties:21]: Yes it is. I'm not sure why the build did not remove those. It seems to have removed others. I'll check.

File: core/camel-core-model/src/main/java/org/apache/camel/model/ThrottleDefinition.java (hunk 5 above)
Comment (orpiske) [reply core/camel-core-model/src/main/java/org/apache/camel/model/ThrottleDefinition.java:65]: Before or after `correlationExpression`. It seems like we enforce a particular order via `@XmlType`:

`@XmlType(propOrder = { "expression", "correlationExpression" })` 

I'm wondering if I should do something like:

`@XmlType(propOrder = { "expression", "mode", "correlationExpression" })`

File: core/camel-core-model/src/main/java/org/apache/camel/model/ThrottleDefinition.java (hunk 6 above)
Comment (orpiske) [reply core/camel-core-model/src/main/java/org/apache/camel/model/ThrottleDefinition.java:182]: Great. I was trying not to be super disruptive, but it makes sense. It's even better if we can get rid of this.

File: core/camel-management-api/src/main/java/org/apache/camel/api/management/mbean/ManagedConcurrentRequestsThrottlerMBean.java (hunk 10 above)
Comment (orpiske) [reply core/camel-management-api/src/main/java/org/apache/camel/api/management/mbean/ManagedConcurrentRequestsThrottlerMBean.java:21]: I am not sure, TBH. 

It seems that the bean works on top of the throttler itself. So, in some way, we would need to rework the code to make the throttler generic. 

I think we can do it, but I believe I might need to tackle the more critical issues first and then this.

File: core/camel-core-reifier/src/main/java/org/apache/camel/reifier/ThrottleReifier.java (hunk 9 above)
Comment (orpiske) [reply core/camel-core-reifier/src/main/java/org/apache/camel/reifier/ThrottleReifier.java:55]: I may need a clarification here. I'll ping you on Zulip / Slack.

File: core/camel-management-api/src/main/java/org/apache/camel/api/management/mbean/ManagedConcurrentRequestsThrottlerMBean.java (hunk 10 above)
Comment (orpiske) [reply core/camel-management-api/src/main/java/org/apache/camel/api/management/mbean/ManagedConcurrentRequestsThrottlerMBean.java]: Done. Also implemented in a way that we could add other types of throttlers in the future.

**Comment by orpiske** [comment]
> I still need to update the documentation and include an entry on the migration guide. I will do this today.

Done.

**Review by davsclaus — APPROVED** [review]

**Pull request merged**